Fix potential race during TcpTransport close

henningandersen opened the pull request:
Fixed two potential causes for leaked threads during tests:
1. When adding a channel to serverChannels, we add it under a monitor
that we do not use when reading from it. This is potentially unsafe if
there is no other happens-before relationship ensuring the safety of
this.
2. Long-shot but if the thread pool was shutdown before entering this
code, we would silently forget about closing server channels so added
assert.

Relates to CI failure issue: #37543

This should be backported to 6.7 and 6.6 too, since MockTcpTransport is not in use (has been removed) in master and 7.0.

henningandersen committed:
Fix potential race during TcpTransport close

Fixed two potential causes for leaked threads during tests:
1. When adding a channel to serverChannels, we add it under a monitor
that we do not use when reading from it. This is potentially unsafe if
there is no other happens-before relationship ensuring the safety of
this.
2. Long-shot but if the thread pool was shutdown before entering this
code, we would silently forget about closing server channels so added
assert.

Relates to CI failure issue: #37543

elasticmachine commented:
Pinging @elastic/es-distributed

henningandersen committed:
Fix potential race during TcpTransport close

Further improve the locking around bind and stop to ensure that once
transport stops and closes serverChannels, no new channels can be added.

Relates to CI failure issue: #37543

DaveCTurner reviewed (approved):
LGTM

original-brownbear commented on server/src/main/java/org/elasticsearch/transport/TcpTransport.java:
```diff
@@ -388,28 +388,35 @@ private InetSocketAddress bindToPort(final String name, final InetAddress hostAd
-                    if (list == null) {
-                        list = new ArrayList<>();
-                        serverChannels.put(name, list);
+        closeLock.readLock().lock();
```
@henningandersen @DaveCTurner shouldn't we acquire the `writeLock` here too to prevent this from running concurrently with the `doStop` (this method is only called from `doStart` anyway, so locking here shouldn't hurt anything since we're not concurrently)?
Otherwise, there isn't really anything that prevents a `put` to the `serverChannels` to run here if the shutdown happened after the above check for the `lifecycle` state passed is there? (which would lead us to still add to `serverChannels` after we're closed already)

henningandersen replied:
@original-brownbear the readLock here should prevent `doStop` from starting to close channels concurrently with this. Notice that lifecycle is moved to `STOPPED` before `doStop` is called and that the lifecycle state is a volatile, so this should be safe I think.

original-brownbear replied:
@henningandersen Yea you're right <= Apparently I shouldn't have answered on a empty stomach.

Still, since this method is only called by `doStart` I'm wondering if the write lock isn't still the better choice since it would allow us to just remove the synchronization on `serverChannels` here and in `doStop` (those are the only two places touching this map) and simplify things a bit?
Just an optional idea though :)

original-brownbear reviewed (approved):
LGTM

henningandersen committed:
Fix potential race during TcpTransport close

Simplify locking around serverChannels a bit.

Relates to CI failure issue: #37543

henningandersen committed:
Fix potential race during TcpTransport close

Simplify locking around serverChannels a bit.

Relates to CI failure issue: #37543
